### ISSUE TITLE
Cluster "master" may remain after entry "server" process dies

### ISSUE THREAD

**Issue by amekkawi** [SocketCluster/socketcluster#211]
If the entry process is killed (-9) or if an unhandled exception is thrown, the process will die but it will not have a chance to tell the cluster "master" process to kill workers.

The port will then remain in use and starting another instance of the server will throw "address in use" errors in the workers.

One solution would be to listen for the "disconnect" event on process within lib/workercluster.js and terminate the cluster master. I'm able to do the same using brokerController but there is no custom code that can be run in the cluster "master" as far as I can tell.

Example:

```javascript
if (cluster.isMaster) {
    if (process.send) {
        process.on('disconnect', function() {
            // terminate the cluster master
        });
    }

    // ...
}
else {
    // ...
}
```

**Comment by amekkawi** [comment]
Alternatively a solution would be to add another option like workerController but for the cluster "master" process.

**Comment by jondubois** [comment]
@amekkawi There is a `'workerExit'` event which you can listen to in the master process - It wasn't in the documentation though so I updated it ;) http://socketcluster.io/#!/docs/api-socketcluster

**Comment by amekkawi** [comment]
Actually this is for when the "server.js" process dies either from a `kill -9` or from an unhandled error.

For example:

```javascript
var SocketCluster = require('socketcluster').SocketCluster;
var socketCluster = new SocketCluster({
    //...
});

setTimeout(function() {
    this.somethingThatDoesntExist();
}, 2000);
```

When the timeout function runs it will throw an error that causes the SocketCluster entry (server.js) process to die. However since it's not a SIGTERM the workers and broker are not told to terminate and will remaining running.

However, the cluster "master" process may eventually throw an error and die if it tries to send a message up to the "server.js" process. I'm assuming the broker would do the same.

The terminology is a little confusing so here is a diagram of the parts:

```
└─┬ SocketCluster entry (i.e. server.js)
  ├── Broker
  └─┬ Cluster "master" (lib/workercluster.js and cluster.isMaster === true)
    └── Cluster "worker" (lib/workercluster.js and cluster.isMaster === false)
```

Note: This is on socketcluster 5.0.4

**Comment by amekkawi** [comment]
Just a note: If the cluster "master" is killed the cluster "worker" will automatically die as well (which is good). I believe they are internally listening for the "disconnect" event on `process` as well, which means the IPC channel of the "master" process was lost.

**Comment by jondubois** [comment]
Oh I see. Yes, that's not ideal.
Ideally all child processes should be shut down if there is an error in server.js.

I guess the solution for now is to add something like this to server.js:

```
process.on('uncaughtException', function (err) {
  console.log(err);
  socketCluster.killWorkers();
  socketCluster.killBrokers();

  process.exit();
});
```

I'll make a patch to make this behavior the default.

**Comment by amekkawi** [comment]
I wouldn't catch the uncaught exceptions just in case someone implementing SocketCluster wants to handle them separately. Since you're exiting right after killing the workers, no other handlers registered after SocketCluster's handler would run.

For example:

```javascript
process.on('uncaughtException', function (err) {
  console.error(err);
  socketCluster.killWorkers(); // This will kill the workercluster master too
  socketCluster.killBrokers();

  process.exit();
});

process.on('uncaughtException', function(err) {
	console.log('Second'); // Would never run
	console.error(err.stack);
});
```

Also if the process is dies by any other method that doesn't allow code to run (out of memory or `kill -9`) then the cluster "master" and brokers would still remain running.

Instead I'd duplicate the behavior of the builtin cluster module (see https://github.com/nodejs/node/blob/master/lib/cluster.js#L553) by listening for the "disconnect" event on `process` in the cluster "master" and broker processes. That event is triggered the the IPC channel is closed/disconnected.

While I'm not sure where this would go for the brokers, an example of the code for the cluster "master" is in my original comment.

For the brokers I'm just inserting that code into the brokerController since it gives me a hook into the process. Since there isn't an equivalent for the cluster "master" I can't add the same hook.

**Comment by jondubois** [comment]
@amekkawi Ah yes, using the 'disconnect' event inside the child processes sounds like the best approach - I did some quick testing and it looks like this approach is working.

**Comment by amekkawi** [comment]
Great! While there is still the issue of the broker, it's possible to address it using the `brokerController` option.

Otherwise it would require an edit to https://github.com/SocketCluster/sc-broker/blob/4323889727f49588cea85247981667ae9f1c6ccb/server.js to also add a 'disconnect' handler. That might make sense since the next time `process.send` is called after a 'disconnect' it will throw an error.

Also thanks for the great library! I was just brought on to a project that uses SocketCluster and with this fix I'm behind it 100% :)

**Comment by jondubois** [comment]
@amekkawi Ok, I fixed it using the 'disconnect' technique you described. It was a good opportunity to cleanup some old code in there which was needlessly complicated; your feedback is much appreciated. I published to npm `v5.0.7`. Note that if you upgrade, you will need to make sure you have the `sc-broker` `v2.2.1` dependency since I had to change the shutdown logic there too (as you assumed).

Let me know if there are any issues.

Also feel free to submit PRs if you like. Raising issues like this is also a great way to contribute.
